pkg/metrics/garbage_collector.go

```diff
@@ -0,0 +1,190 @@
+	logrus.Debugf("MetricGarbageCollector Start")
+
+	clusterLister := context.Management.Clusters("").Controller().Lister()
+	nodeLister := context.Management.Nodes("").Controller().Lister()
+	endpointLister := context.Core.Endpoints(settings.Namespace.Get()).Controller().Lister()
+	isClusterMode := settings.Namespace.Get() != "" && settings.PeerServices.Get() != ""
+
+	// Existing resources
+	observedResourceNames := map[string]bool{}
+	// Exisiting Metrics Labels
+	observedLabelsMap := map[string]map[interface{}][]map[string]string{}
+
+	// Get Clusters
+	clusters, err := clusterLister.List("", labels.Everything())
+	if err != nil {
+		logrus.Errorf("MetricGarbageCollector failed to list clusters: %s", err)
+		return
+	}
+	for _, cluster := range clusters {
+		if _, ok := observedResourceNames[cluster.Name]; !ok {
+			observedResourceNames[cluster.Name] = true
+		}
+	}
+	// Get Nodes
+	nodes, err := nodeLister.List("", labels.Everything())
+	if err != nil {
+		logrus.Errorf("MetricGarbageCollector failed to list nodes: %s", err)
+	}
+	for _, node := range nodes {
+		if _, ok := observedResourceNames[node.Namespace+":"+node.Name]; !ok {
+			observedResourceNames[node.Namespace+":"+node.Name] = true
+		}
+	}
+	// Get Endpoints
+	endpoints, err := endpointLister.List(settings.Namespace.Get(), labels.Everything())
+	countOfRancherServer := 0
+	if isClusterMode {
+		for _, svc := range strings.Split(settings.PeerServices.Get(), ",") {
+			for _, e := range endpoints {
+				// If Endpoint is associated with PeerServices service
+				if e.Name == strings.TrimSpace(svc) {
+					for _, subset := range e.Subsets {
+						for _, addr := range subset.Addresses {
+							observedResourceNames[addr.IP] = true
+							countOfRancherServer++
+						}
+					}
+				}
+			}
+		}
+	}
+	logrus.Debugf("MetricGarbageCollector saw %d clusters", len(clusters))
+	logrus.Debugf("MetricGarbageCollector saw %d nodes", len(nodes))
+	logrus.Debugf("MetricGarbageCollector saw %d rancher servers", countOfRancherServer)
+
+	observedClientKey := BuildObservedLabelMaps(GCTargetMetricsByNameForClientKey, "clientkey", observedLabelsMap)
+	observedPeer := BuildObservedLabelMaps(GCTargetMetricsByIPForPeer, "peer", observedLabelsMap)
+	observedCluster := BuildObservedLabelMaps([]interface{}{ClusterOwner}, "cluster", observedLabelsMap)
+	logrus.Debugf("MetricGarbageCollector saw %d clientkey of metrics", observedClientKey)
+	logrus.Debugf("MetricGarbageCollector saw %d peer of metrics", observedPeer)
+	logrus.Debugf("MetricGarbageCollector saw %d cluster of metrics", observedCluster)
+
+	removedCount := RemoveMetricsForDeletedResource(observedLabelsMap, observedResourceNames)
+	logrus.Debugf("MetricGarbageCollector removed %d metrics", removedCount)
+
+	logrus.Debugf("MetricGarbageCollector Finished")
+}
+
+func BuildObservedLabelMaps(collectors []interface{}, targetLabel string, observedLabels map[string]map[interface{}][]map[string]string) int {
```

Comment:
This may require breaking this out into separate functions to avoid using the empty interface